org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/ProjectCachePlunger.java

```diff
@@ -62,25 +62,36 @@ public void register(MavenProject project, Key cacheKey) {
     }
     final Set<File> affectedProjects = new HashSet<>();
 
-    for(Key cacheKey : projectKeys.removeAll(pom)) {
-      keyProjects.remove(cacheKey, pom);
+    for(Key cacheKey : removeAll(projectKeys, pom)) {
+      remove(keyProjects, cacheKey, pom);
       if(forceDependencyUpdate && RepositoryPolicy.UPDATE_POLICY_ALWAYS.equals(session.getUpdatePolicy())
           && session.getCache().get(session, cacheKey) == null) {
         session.getCache().put(session, cacheKey, Boolean.TRUE);
-        for(File affectedPom : keyProjects.removeAll(cacheKey)) {
+        for(File affectedPom : removeAll(keyProjects, cacheKey)) {
           affectedProjects.add(affectedPom);
-          projectKeys.remove(affectedPom, cacheKey);
+          remove(projectKeys, affectedPom, cacheKey);
         }
       }
       if(!keyProjects.containsKey(cacheKey)) {
         flush(cacheKey);
         log.debug("Flushed cache entry for {}", cacheKey);
       }
     }
-
     return affectedProjects;
   }
 
+  private static <K, V> Set<V> removeAll(Map<K, Set<V>> map, K key) {
```

Comment:
> > especially in one use case of Multimaps
> 
> Reading though the javadoc Multimap seems to be quite easlie replaced by getOrDefault / computeIfAbsent or even Collectors.groupBy, could you explain what has made it complex here?

In most cases you're right. When mentioning 'problems' I was mainly referring to this and the following method.
This is not really bad, but a bit longer than before.